market_making_examples/restingLimitMaker/index.ts

```diff
@@ -0,0 +1,484 @@
+	const spreadDiffBps = minSpreadBps + (maxSpreadBps - minSpreadBps) * (currLeverage / targetLeverage);
+
+	if (currentBasePosition == 0) {
+		// no position, quote tightest
+		return {
+			baseBidPrice: dlobMidPrice * (1 - spreadDiffBps / 10000.0),
+			baseAskPrice: dlobMidPrice * (1 + spreadDiffBps / 10000.0),
+		};
+	} else if (currentBasePosition > 0) {
+		// currently long quote wider on the bid
+		return {
+			baseBidPrice: dlobMidPrice * (1 - spreadDiffBps / 10000.0),
+			baseAskPrice: dlobMidPrice * (1 + minSpreadBps / 10000.0),
+		};
+	} else if (currentBasePosition < 0) {
+		// currently short quote wider on the ask
+		return {
+			baseBidPrice: dlobMidPrice * (1 - minSpreadBps / 10000.0),
+			baseAskPrice: dlobMidPrice * (1 + spreadDiffBps / 10000.0),
+		};
+	} else {
+		throw new Error("Invalid position");
+	}
+};
+
+
+const needUpdateOrders = (expectedOpenSize: BN, openBids: BN, openAsks: BN, dlobMidPrice: number, myBaseBid: number, myBaseAsk: number, vBid: number, vAsk: number): boolean => {
+	// update orders if any orders have been completely filled
+	if (!openBids.abs().eq(expectedOpenSize)) {
+		logger.info("Requoting due to filled bids");
+		return true;
+	}
+	if (!openAsks.abs().eq(expectedOpenSize)) {
+		logger.info("Requoting due to filled ask order");
+		return true;
+	}
+
+	// update orders if the dlob midprice is less than min spread from our base bid/ask
+	const bidThreshold = myBaseBid * (1 + minSpreadBps / 2 / 10000.0);
+	if (dlobMidPrice < bidThreshold) {
+		logger.info("Requoting due to dlobMidPrice approaching bid");
+		return true;
+	}
+	const askThreshold = myBaseAsk * (1 - minSpreadBps / 2 / 10000.0);
+	if (dlobMidPrice > askThreshold) {
+		logger.info("Requoting due to dlobMidPrice approaching ask");
+		return true;
+	}
+
+	// dont want vamm to cross us
+	if (vBid > askThreshold) {
+		logger.info(`Requoting due to vBid (${vBid}) approaching askThres (${askThreshold})`);
+		return true;
+	}
+	if (vAsk < bidThreshold) {
+		logger.info(`Requoting due to vAsk (${vAsk}) approaching bidThres (${bidThreshold})`);
+		return true;
+	}
+
+
+	return false;
+};
+
+/**
+ * Places orders on both sides of the book, starting at base*price, and increasing by ticksBetweenOrders
+ * until ordersPerSide are placed.
+ * @param driftClient
+ * @param baseBidPrice
+ * @param baseAskPrice
+ */
+const updateOrders = async (driftClient: DriftClient, baseBidPrice: number, baseAskPrice: number) => {
+	// load open bids and asks from chain state
+	const openBids: Array<Order> = [];
+	const openAsks: Array<Order> = [];
+	for (const order of driftClient.getUserAccount()!.orders) {
+		if (isVariant(order.status, "init")) {
+			continue;
+		}
+
+		if (isVariant(order.direction, "long")) {
+			openBids.push(order);
+		} else {
+			openAsks.push(order);
+		}
+	}
+
+	// reduce, resuse, recycle open order slots
+	const orderExpireTs = new BN((Date.now()) / 1000 + ORDER_TIF_S); // new order will expire in 30s
```

Comment:
jarry's phoenix code had a way to track on-chain time, we could maybe add that to sdk